tensorflow_probability/python/stats/kendalls_tau.py

```diff
@@ -0,0 +1,166 @@
+  """Computes Kendall's Tau for two ordered lists.
+
+  Kendall's Tau measures the correlation between ordinal rankings. This
+  implementation is similar to the one used in scipy.stats.kendalltau.
+  Args:
+    y_true: A tensor that provides a true ordinal ranking of N items.
+    y_pred: A presumably model provided ordering of the same N items:
+
+  Returns:
+    Kendell's Tau, the 1945 tau-b formulation that ignores ordering of
+    ties, as a scalar Tensor.
+  """
+  y_true = tf.reshape(y_true, [-1])
+  y_pred = tf.reshape(y_pred, [-1])
+  y_pred.shape.assert_is_compatible_with(y_true.shape)
+  if tf.equal(tf.size(y_true), 0) or tf.equal(tf.size(y_pred), 0):
+    warnings.warn("y_true and y_pred tensors are not the same size.")
+    return np.nan
+  perm = tf.argsort(y_true)
+  n = tf.shape(perm)[0]
+  if tf.less(n, 2):
+    warnings.warn("Scalar tensors have no defined ordering.")
+    return np.nan
+
+  left = 0
+  # scan for ties, and for each range of ties do a argsort on
+  # the y_pred value. (TF has no lexicographical sorting, although
+  # jax can sort complex number lexicographically. Hmm.)
+  lexi = tf.TensorArray(tf.int32, size=n)
+  for i in tf.range(n):
+    lexi = lexi.write(i, perm[i])
+  for right in tf.range(1, n):
+    ytruelr = tf.gather(y_true, tf.gather(perm, [left, right]))
+    if tf.not_equal(ytruelr[0], ytruelr[1]):
+      sub = perm[left:right]
+      subperm = tf.argsort(tf.gather(y_pred, sub))
+      lexi = lexi.scatter(tf.range(left, right), tf.gather(sub, subperm))
+      left = right
+  sub = perm[left:n]
+  subperm = tf.argsort(tf.gather(y_pred, perm[left:n]))
+  lexi.scatter(tf.range(left, n), tf.gather(sub, subperm))
+
+  # This code follows roughly along with scipy/stats/stats.py v. 0.15.1
+  # compute joint ties
+  first = 0
+  t = 0
+  for i in tf.range(1, n):
+    permfirsti = lexi.gather([first, i])
+    y_truefirsti = tf.gather(y_true, permfirsti)
+    y_predfirsti = tf.gather(y_pred, permfirsti)
+    if y_truefirsti[0] != y_truefirsti[1] or y_predfirsti[0] != y_predfirsti[1]:
+      t += ((i - first) * (i - first - 1)) // 2
+      first = i
+  t += ((n - first) * (n - first - 1)) // 2
+
+  # compute ties in y_true
+  first = 0
+  u = 0
+  for i in tf.range(1, n):
+    y_truefirsti = tf.gather(y_true, lexi.gather([first, i]))
+    if y_truefirsti[0] != y_truefirsti[1]:
+      u += ((i - first) * (i - first - 1)) // 2
+      first = i
+  u += ((n - first) * (n - first - 1)) // 2
+
+  # count exchanges
+  exchanges, newperm = _iterative_mergesort(y_pred, lexi)
+  # compute ties in y_pred after mergesort with counting
+  first = 0
+  v = 0
+  for i in tf.range(1, n):
+    y_predfirsti = tf.gather(y_pred, tf.gather(newperm, [first, i]))
+    if y_predfirsti[0] != y_predfirsti[1]:
+      v += ((i - first) * (i - first - 1)) // 2
+      first = i
+  v += ((n - first) * (n - first - 1)) // 2
+
+  tot = (n * (n - 1)) // 2
+  if tf.equal(tot, u) or tf.equal(tot, v):
```

Comment:
As above---prefer raising an error (`assert_util.assert_none_equal(tot, y)`) to returning `NaN`.